File: src/data/codes/linear-regression.m
clc;
clear;
close all;

%% =========================
%  SETTINGS
% ==========================
useSynthetic = true;   % true = use built-in synthetic dataset
csvFile = 'bodyfat_data.csv';  % used only if useSynthetic = false

%% =========================
%  LOAD OR CREATE DATA
% ==========================
if useSynthetic
    rng(1);   % for repeatable results
    n = 100;

    Age = randi([18 60], n, 1);
    DailyCalorieIntake = randi([1800 3500], n, 1);
    SleepHours = 5 + 4*rand(n,1);              % between 5 and 9
    WorkoutSessions = randi([0 6], n, 1);

    % Synthetic target with some noise
    BodyFat = 0.25*Age + 0.004*DailyCalorieIntake - 1.8*SleepHours ...
              - 1.5*WorkoutSessions + randn(n,1)*2 + 8;

    data = table(Age, DailyCalorieIntake, SleepHours, WorkoutSessions, BodyFat);
else
    data = readtable(csvFile);

    % Example expected column names:
    % Age, DailyCalorieIntake, SleepHours, WorkoutSessions, BodyFat
end

disp('First 5 rows of dataset:');
disp(head(data));

%% =========================
%  PREPROCESSING
% ==========================
% Remove missing values
data = rmmissing(data);

% Remove duplicate rows
data = unique(data);

% Separate input and target
X = [data.Age, data.DailyCalorieIntake, data.SleepHours, data.WorkoutSessions];
y = data.BodyFat;

% Normalize features
X = normalize(X);

%% =========================
%  TRAIN-TEST SPLIT
% ==========================
cv = cvpartition(size(X,1), 'HoldOut', 0.2);
X_train = X(training(cv), :);
y_train = y(training(cv), :);

X_test = X(test(cv), :);
y_test = y(test(cv), :);

%% =========================
%  MODEL BUILDING
% ==========================
model = fitlm(X_train, y_train);

disp('Linear Regression Model:');
disp(model);

%% =========================
%  PREDICTION
% ==========================
y_pred = predict(model, X_test);

%% =========================
%  EVALUATION
% ==========================
MAE = mean(abs(y_test - y_pred));
MSE = mean((y_test - y_pred).^2);
RMSE = sqrt(MSE);
R2 = 1 - sum((y_test - y_pred).^2) / sum((y_test - mean(y_test)).^2);

fprintf('\nEvaluation Metrics:\n');
fprintf('MAE  = %.4f\n', MAE);
fprintf('MSE  = %.4f\n', MSE);
fprintf('RMSE = %.4f\n', RMSE);
fprintf('R^2  = %.4f\n', R2);

%% =========================
%  VISUALIZATION
% ==========================
figure;
scatter(y_test, y_pred, 'filled');
xlabel('Actual Body Fat');
ylabel('Predicted Body Fat');
title('Actual vs Predicted Body Fat');
grid on;

figure;
plot(y_test, 'b-o', 'LineWidth', 1.5);
hold on;
plot(y_pred, 'r-*', 'LineWidth', 1.5);
legend('Actual', 'Predicted');
title('Actual and Predicted Values');
xlabel('Test Sample');
ylabel('Body Fat');
grid on;

File: src/data/codes/linear-regression.py
"""
Linear regression — numpy / sklearn
# pip install numpy pandas scikit-learn matplotlib
"""
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import matplotlib.pyplot as plt

# Settings
USE_SYNTHETIC = True
CSV_FILE = "bodyfat_data.csv"
RNG = np.random.default_rng(1)

if USE_SYNTHETIC:
    n = 100
    age = RNG.integers(18, 61, size=n)
    daily_calorie = RNG.integers(1800, 3501, size=n)
    sleep_hours = RNG.uniform(5, 9, size=n)
    workout_sessions = RNG.integers(0, 7, size=n)
    body_fat = (
        0.25 * age
        + 0.004 * daily_calorie
        - 1.8 * sleep_hours
        - 1.5 * workout_sessions
        + RNG.normal(0, 2, size=n)
        + 8
    )
    data = pd.DataFrame(
        {
            "Age": age,
            "DailyCalorieIntake": daily_calorie,
            "SleepHours": sleep_hours,
            "WorkoutSessions": workout_sessions,
            "BodyFat": body_fat,
        }
    )
else:
    data = pd.read_csv(CSV_FILE)

print(data.head())

data = data.dropna().drop_duplicates()
X = data[["Age", "DailyCalorieIntake", "SleepHours", "WorkoutSessions"]].values
y = data["BodyFat"].values
scaler = StandardScaler()
X = scaler.fit_transform(X)

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

model = LinearRegression().fit(X_train, y_train)
y_pred = model.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred)

print(f"MAE  = {mae:.4f}")
print(f"MSE  = {mse:.4f}")
print(f"RMSE = {rmse:.4f}")
print(f"R^2  = {r2:.4f}")

fig, ax = plt.subplots()
ax.scatter(y_test, y_pred, alpha=0.7)
ax.set_xlabel("Actual Body Fat")
ax.set_ylabel("Predicted Body Fat")
ax.set_title("Actual vs Predicted Body Fat")
ax.grid(True)

fig2, ax2 = plt.subplots()
idx = np.arange(len(y_test))
ax2.plot(idx, y_test, "b-o", label="Actual", linewidth=1.5)
ax2.plot(idx, y_pred, "r-*", label="Predicted", linewidth=1.5)
ax2.legend()
ax2.set_title("Actual and Predicted Values")
ax2.set_xlabel("Test Sample")
ax2.set_ylabel("Body Fat")
ax2.grid(True)
plt.show()
